### ISSUE TITLE
debug exe in user mode

### ISSUE THREAD

**Issue by xSanx** [HyperDbg/HyperDbg#92]
hi guys! using `.load` I can load and debug kernel modules, and how can I run and debug a regular .exe file?

**Comment by SinaKarvandi** [comment]
Hi,
Debugging user-mode applications with the mean of classic debuggers in local mode is not yet supported in HyperDbg. We once tried to implement it but it failed due to the complexity of designing an OS-independent mechanism to support user-mode applications without halting the entire kernel. So, we left to the future version. But, debugging user-mode applications is fully supported in the [Debugger Mode](https://docs.hyperdbg.org/using-hyperdbg/prerequisites/operation-modes#debugger-mode).

If you are connected to Debugger Mode, you can debug everything including the user-mode applications. Please follow the steps here: https://docs.hyperdbg.org/getting-started/attach-to-hyperdbg/debug#connect-to-debuggee-debugger-mode

You can switch to different processes using the [.thread](https://docs.hyperdbg.org/commands/meta-commands/.thread) and the [.process](https://docs.hyperdbg.org/commands/meta-commands/.process) commands:

Here is an example of debugging a user-mode application:
https://docs.hyperdbg.org/using-hyperdbg/examples/basics/switching-to-a-specific-process-or-thread

**Comment by DarkerSquirrel** [comment]
This is a good question , Sina, I intended to ask that too.

Load/Attach to a process is the first thing a user space debugger does  and many users(myself included) were baffled first
time when we encountered Hyperdbg and not finding it.

I am v curious of how do you envision the future workflow a HyperDbg regarding the debugging of user processes.
Is HyperDbg supposed to work together with an user space debugger like x64dbg, providing extra functionalities(like unlimited hwbps,etc) or HyperDbg will eventually will support all the functionalities that x64dbg is providing now?

One thing I am thinking of is for example running a condition(triggered by an event) in user space also , not only in VMX root mode. Also would it be possible for that condition to be just not a series of opcodes , but an entry function of a dll/.sys or something that would allow coding in C++ instead of assembler?

I think more questions could come to mind , but I want to get more deeper in the docs before asking them.

Thanks

**Comment by SinaKarvandi** [comment]
Hi,

Thanks for asking the questions. This answer represents my thoughts about HyperDbg, and of course, other developers or possible future developers might not agree with me, and I don't wanna create a general roadmap for everyone in this project.

I believe that x64dbg is by far the best debugger in debugging user-mode applications. HyperDbg is not designed for this purpose, and it's by no means comparable with x64dbg.

Nowadays, it's substantially harder to investigate modern packed/protected programs, and the APIs that the operating systems like Windows provide for debugging is not enough. That's why we designed HyperDbg. 

In my opinion, I see HyperDbg with many hypervisor-based and new processor-based techniques that will create a core for debugging and brings lots of new and useful methods for reverse engineering. Of course, other debuggers like x64dbg can use these features (and we love to integrate with those debuggers).

> HyperDbg will eventually will support all the functionalities that x64dbg is providing now?

Yes, but not the way that Windows implement those APIs or by the means of classic debuggers.
We never used any debugging API, so we will implement all the methods like attaching or stepping through the user-mode instructions with OS independency in mind, and the use of virtualization technology. If we got a chance to add our APIs for attaching and stepping to the x64dbg and synchronize instructions and memory panels with x64dbg, then that would be great. Because the GUI and the features that are implemented in x64dbg are unique and we neither want nor can implement these things again.

> Is it possible now in HyperDbg to get a callstack / Loaded modules when a bp is hit?

Yep but soon, I think @aleeamini is working on the call stack right now.
